### PR TITLE
UI: [VAULT-22908] Update kv object editor, add max number of messages reached modal, small improvements

### PR DESCRIPTION
**Pull request by kiannaquach**
**Description**
- Update kv object editor to show only one row

Screenshot 2024-01-17 at 12 11 16 PM

The backend expects the link attribute with the following data structure:
```

  {
    link: {
      'hello world': 'www.blah.com'
    }
  }
```
Since this follows a similar structure to the kv-object-editor, we decided to utilize this component.

- Added maximum message limit reached modal:

Screenshot 2024-01-18 at 12 03 42 PM

- Updated banner colors to be `highlight` and to be dismissable based on previous feedback
Screenshot 2024-01-19 at 9 10 21 AM

### CODE REVIEW

File: ui/app/models/config-ui/message.js [hunk 1]
```diff
@@ -11,10 +11,21 @@ import { withFormFields } from 'vault/decorators/model-form-fields';
...
+  link: [
+    {
+      validator(model) {
+        if (!model?.link) return true;
+        const linkTitle = Object.keys(model.link).length === 1;
+        const linkHref = Object.values(model.link);
+        return linkTitle && !!linkHref[0];
+      },
+      message: 'Link href is required.',
+    },
```
Comment (kiannaquach) [inline ui/app/models/config-ui/message.js:23]: Link is an optional form field, however, if a user enters a link title, they must also enter an href.

File: ui/tests/acceptance/mfa-login-test.js [hunk 2]
```diff
@@ -35,7 +35,7 @@ module('Acceptance | mfa-login', function (hooks) {
   });
 
   const login = async (user) => {
-    await visit('/vault/auth');
+    await visit('/vault/auth?with=token');
```
Comment (kiannaquach) [inline ui/tests/acceptance/mfa-login-test.js:38]: MFA tests were failing since the auth form had different tabs and it was defaulting to the oidc tab, so this enforces that it will go to the auth form with the dropdown of auth methods.

Comment (hellobontempo) [reply ui/tests/acceptance/mfa-login-test.js:38]: 🤔 This means something wasn't clearing out in a previous test correctly. I wonder what changed!

File: ui/lib/core/addon/components/form-field.hbs [hunk 3]
```diff
@@ -133,12 +133,15 @@
       @value={{get @model this.valuePath}}
       @onChange={{this.setAndBroadcast}}
       @label={{this.labelString}}
-      @labelClass="title {{if (eq @mode 'create') 'is-5' 'is-4'}}"
+      @labelClass={{unless @attr.options.shouldOnlyShowSingleRow (if (eq @mode "create") "title is-5" "title is-4")}}
```
Comment (hellobontempo) [inline ui/lib/core/addon/components/form-field.hbs:136]: I changed this in my most recent PR, I recommend merging in that change so that this instead would be controlled by `attr.options.isSectionHeader`
```js
@labelClass={{if @attr.options.isSectionHeader "title is-4" "is-label"}}

```

This way the section header/form-section styling isn't reliant on `onlyShowSingleRow` which it could be we want multiple rows _and_  section header styling

File: ui/lib/core/addon/components/form-field.hbs [hunk 4]
```diff
@@ -133,12 +133,15 @@
...
       @helpText={{@attr.options.helpText}}
       @subText={{@attr.options.subText}}
       @onKeyUp={{this.handleKeyUp}}
+      @keyPlaceholder={{@attr.options.keyPlaceholder}}
+      @valuePlaceholder={{@attr.options.valuePlaceholder}}
       @validationError={{this.validationError}}
-      class="form-section"
+      @shouldOnlyShowSingleRow={{@attr.options.shouldOnlyShowSingleRow}}
+      class={{unless @attr.options.shouldOnlyShowSingleRow "form-section"}}
```
Comment (hellobontempo) [inline ui/lib/core/addon/components/form-field.hbs:144]: this was also changed in my PR to the following (`isSectionHeader` is already an attr option we use for search select styling)
```js
      class={{if @attr.options.isSectionHeader "form-section"}}

File: ui/app/models/config-ui/message.js [hunk 5]
```diff
@@ -81,11 +92,14 @@ export default class MessageModel extends Model {
   startTime;
   @attr('dateTimeLocal', { editType: 'yield', label: 'Message expires' }) endTime;
 
-  // the api returns link as an object with title and href as keys, but we separate the link key/values into
-  // different attributes to easily show link title and href fields on the create form. In our serializer,
-  // we send the link attribute in to the correct format (as an object) to the server.
-  @attr('string') linkTitle;
-  @attr('string') linkHref;
+  @attr('object', {
```
Comment (hellobontempo) [inline ui/app/models/config-ui/message.js:95]: Making this an object is much more intuitive! Nice improvement

File: ui/app/models/config-ui/message.js [hunk 6]
```diff
@@ -81,11 +92,14 @@ export default class MessageModel extends Model {
...
+    editType: 'kv',
+    keyPlaceholder: 'Display text (e.g. Learn more)',
+    valuePlaceholder: 'Paste URL (e.g. www.learnmore.com)',
+    label: 'Link (optional)',
+    shouldOnlyShowSingleRow: true,
```
Comment (hellobontempo) [inline ui/app/models/config-ui/message.js:100]: I'd rename this `isSingleRow`

**Review by hellobontempo — COMMENTED** [review]
Great work! Just a couple suggestions 😄

File: ui/lib/core/addon/components/form-field.hbs (hunk 3 above)
Comment (hellobontempo) [reply ui/lib/core/addon/components/form-field.hbs:136]: Lol I forgot I haven't merged my PR...you can just revert these changes and you'll be good!

File: ui/app/models/config-ui/message.js [hunk 7]
```diff
@@ -11,10 +11,21 @@ import { withFormFields } from 'vault/decorators/model-form-fields';
 const validations = {
   title: [{ type: 'presence', message: 'Title is required.' }],
   message: [{ type: 'presence', message: 'Message is required.' }],
+  link: [
+    {
+      validator(model) {
+        if (!model?.link) return true;
+        const linkTitle = Object.keys(model.link).length === 1;
+        const linkHref = Object.values(model.link);
+        return linkTitle && !!linkHref[0];
```
Comment (hellobontempo) [inline ui/app/models/config-ui/message.js:20]: You could use descructuring here
```suggestion
        const [linkHref] = Object.values(model.link);
        return linkTitle && !!linkHref;

File: ui/lib/core/addon/components/form-field.hbs (hunk 3 above)
Comment (kiannaquach) [reply ui/lib/core/addon/components/form-field.hbs:136]: Thanks Claire!

File: ui/app/templates/vault/cluster.hbs [hunk 8]
```diff
@@ -3,37 +3,37 @@
...
-    {{/each}}
-  {{/if}}
-{{/if}}
+{{#each this.customMessages.bannerMessages as |bannerMessage|}}
+  <Hds::Alert @type="inline" @color="warning" data-test-alert={{bannerMessage.id}} as |A|>
+    <A.Title data-test-alert-title={{bannerMessage.id}}>{{bannerMessage.title}}</A.Title>
+    <A.Description data-test-alert-description={{bannerMessage.id}}>
+      {{bannerMessage.message}}
+      {{#each-in bannerMessage.link as |title href|}}
+        {{#if (and title href)}}
```
Comment (hellobontempo) [inline ui/app/templates/vault/cluster.hbs:12]: I don't believe this conditional is necessary since there will be nothing to iterate over if the object is empty

File: ui/lib/config-ui/addon/components/messages/page/create-and-edit-message-form.hbs [hunk 9]
```diff
@@ -18,43 +18,15 @@
...
-          class="has-bottom-margin-m"
-        >
-          <Messages::MessageExpirationDateForm @message={{@message}} @attr={{attr}} />
-        </FormField>
-      {{/if}}
+      <FormField
+        @attr={{attr}}
+        @model={{@message}}
+        @modelValidations={{this.modelValidations}}
+        @mode={{if @message.isNew "create" "edit"}}
```
[the remark this answers is not in the document]
Comment (kiannaquach) [reply ui/lib/config-ui/addon/components/messages/page/create-and-edit-message-form.hbs:25]: Good catch! I forgot to remove this :)

File: ui/app/templates/vault/cluster.hbs (hunk 8 above)
Comment (kiannaquach) [reply ui/app/templates/vault/cluster.hbs:12]: That's what I was thinking too, but ran into an error where I was passing undefined to href for <Hds::Link::Inline />

`<Hds::Link::Inline @href={{href}}>{{title}}</Hds::Link::Inline>`

File: ui/lib/config-ui/addon/components/messages/page/details.hbs [hunk 10]
```diff
@@ -43,6 +43,10 @@
         (date-format (get @message attr.name) "MMM d, yyyy hh:mm aaa" withTimeZone=true)
       }}
     />
+  {{else if (eq attr.name "link")}}
+    {{#each-in @message.link as |title|}}
+      <InfoTableRow @label="Link" @value={{if title true false}} />
+    {{/each-in}}
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/details.hbs:49]: Does this render `true` if a link exists but doesn't actually display the link content? It seems like rendering the title and href would be useful 🤔

Comment (kiannaquach) [reply ui/lib/config-ui/addon/components/messages/page/details.hbs:49]: Screenshot 2024-01-18 at 10 06 16 AM
It'll render "Yes" or "No" based on whether or not a link exists. I'll follow up with design on how it should look!

File: ui/tests/acceptance/mfa-login-test.js (hunk 2 above)
Comment (kiannaquach) [reply ui/tests/acceptance/mfa-login-test.js:38]: Hmm I'm curious too! I might explore it more after finishing up these tickets :D

File: ui/lib/config-ui/addon/components/messages/page/details.hbs (hunk 10 above)
Comment (kiannaquach) [reply ui/lib/config-ui/addon/components/messages/page/details.hbs:49]: Edit: spoke with design and we decided to add the link!
Screenshot 2024-01-18 at 11 22 21 AM

File: ui/app/templates/vault/cluster.hbs (hunk 8 above)
Comment (hellobontempo) [reply ui/app/templates/vault/cluster.hbs:12]: Ah yeah - I wondered if the HDS component would error

Another option is wrapping the `{{#each-in}}` in a conditional to avoid iterating all together (in my PR which I'm still working on tests for 😭 I moved the `is-empty-value` template helper to the core addon
```hbs
{{#unless (is-empty-value bannerMessage.link}}
 {{#each-in bannerMessage.link as |title href|}}

Comment (kiannaquach) [reply ui/app/templates/vault/cluster.hbs]: Awesome I will try that!

File: ui/app/serializers/config-ui/message.js [hunk 11]
```diff
@@ -35,12 +32,8 @@ export default class MessageSerializer extends ApplicationSerializer {
...
     const json = super.serialize(...arguments);
     json.message = encodeString(json.message);
-    json.link = {
-      title: json?.link_title || '',
-      href: json?.link_href || '',
-    };
-    delete json?.link_title;
-    delete json?.link_href;
+    const [linkTitle] = Object.keys(json.link);
+    json.link = isEmpty(linkTitle) ? null : json.link;
```
Comment (kiannaquach) [inline ui/app/serializers/config-ui/message.js:36]: Adding these checks since the json.link could be set like `{"": ""}`. If a user edits the link form field, the fields save the object with a key and value as empty strings.

File: ui/tests/integration/components/config-ui/messages/page/list-test.js [hunk 12]
```diff
@@ -96,4 +95,44 @@ module('Integration | Component | messages/page/list', function (hooks) {
       assert.dom(`[data-linked-block-title="${message.id}"]`).hasText(message.title);
     }
   });
+
+  test('it should show max message warning modal', async function (assert) {
```
Comment (kiannaquach) [inline ui/tests/integration/components/config-ui/messages/page/list-test.js:99]: This test is causing pagination accessibility errors. Open to thoughts on how to refine this test so it doesn't cause these accessibility errors. I checked in the ui with the axedevtools and didn't see any accessibility errors, so I think it's how I set up the meta tag here.

File: ui/app/models/config-ui/message.js [hunk 13]
```diff
@@ -11,10 +11,21 @@ import { withFormFields } from 'vault/decorators/model-form-fields';
 const validations = {
   title: [{ type: 'presence', message: 'Title is required.' }],
   message: [{ type: 'presence', message: 'Message is required.' }],
+  link: [
+    {
+      validator(model) {
+        const linkTitle = Object.keys(model.link);
+        const [linkHref] = Object.values(model.link);
+        if (!model?.link || linkTitle.length < 1 || linkTitle[0] === '') return true;
```
[the remark this answers is not in the document]
Comment (hellobontempo) [reply ui/app/models/config-ui/message.js:19]: If link is undefined then I think `Object.keys(model.link);` will fail because that's `Object.keys(undefined);` 

So I think `!model?.link` still needs to be the first conditional, then the second part could be

```js
 validator(model) {
        if ((!model?.link) return true
        const [linkTitle] = Object.keys(model.link); // linkTitle could be undefined or an empty string, both are falsy
        const [linkHref] = Object.values(model.link);
        if (!linkTitle) return true 
        return !!linkHref;
},

[the remark this answers is not in the document]
Comment (hellobontempo) [reply ui/app/models/config-ui/message.js:19]: Here's that demo'd in the console to make sure I'm thinking of the logic correctly
Screenshot 2024-01-18 at 2 25 01 PM

File: ui/tests/integration/components/config-ui/messages/page/list-test.js (hunk 12 above)
Comment (kiannaquach) [reply ui/tests/integration/components/config-ui/messages/page/list-test.js:99]: Edit: I was able to avoid these accessibility errors when I see pageSize to 100. I noticed that accessibility errors don't occur for disabled prev / next buttons.

File: ui/app/serializers/config-ui/message.js (hunk 11 above)
Comment (hellobontempo) [reply ui/app/serializers/config-ui/message.js:36]: This seems like something the component shouldn't allow 🤔 Likely outside the scope of this PR. Or does it seem like the object transform should handle this case? There are other parts of the app where this has been an issue, I think. 

The `serialize` method in transforms/object.js could be updated so if the object is just empty strings, it returns`{}` 
```js
 serialize(value) {
    if (typeOf(value) !== 'object') {
      return {};
    } else {
      return JSON.stringify(value) === '{"":""}' ? {} : value
    }
  }
```
Also an empty string is falsy, so there's no need to import `isEmpty`, the check could be 
```js
!linkTitle ? null : json.link;

File: ui/app/services/custom-messages.js [hunk 14]
```diff
@@ -3,15 +3,17 @@
...
 import { tracked } from '@glimmer/tracking';
-
+import { TrackedObject } from 'tracked-built-ins';
 export default class CustomMessagesService extends Service {
   @service store;
   @service namespace;
   @service auth;
   @tracked messages = [];
   @tracked showMessageModal = true;
+  bannerDismissObject = new TrackedObject();
```
Comment (hellobontempo) [inline ui/app/services/custom-messages.js:16]: This manages the state for each banner? The logic here is a little hard to follow because this name implies that we'd set to `true` if a banner was dismissed, but below it's set to false. 

Could we rename it to `bannerState` or switch the logic if you want this to be `bannerDismissState` so that `true` means a banner has been dismissed and therefore hidden? If this service is converted to typescript then we'd know it was an object, but also seeing `new TrackedObject();` gives us this information so I don't think that needs to be part of the variable name.

Changing the name also makes the template logic easier to follow because it reads as
```hbs
{{#if (get this.customMessages.bannerState bannerMessage.id)}}
// banner state is true so render
{{/if}}
```
OR 
```hbs
{{#unless (get this.customMessages.bannerDismissState bannerMessage.id)}}
// banner dismissed is false so render
{{/unless}

File: ui/app/templates/vault/cluster.hbs [hunk 15]
```diff
@@ -3,37 +3,45 @@
...
-          {{bannerMessage.message}}
-        </A.Description>
-        <A.Description class="has-top-margin-xs">
-          {{! TODO: VAULT-22908 display links when api is updated to { link: { 'learn': 'www.learn.com'} } }}
-        </A.Description>
-      </Hds::Alert>
-    {{/each}}
+{{#each this.customMessages.bannerMessages as |bannerMessage|}}
+  {{#if (get this.customMessages.bannerDismissObject bannerMessage.id)}}
+    <Hds::Alert
```
Comment (hellobontempo) [inline ui/app/templates/vault/cluster.hbs:8]: looking at the screenshot do we want a little margin below each banner separating them slightly?

File: ui/app/templates/vault/cluster.hbs [hunk 16]
```diff
@@ -3,37 +3,45 @@
...
+      @type="inline"
+      @color="highlight"
+      data-test-alert={{bannerMessage.id}}
```
Comment (hellobontempo) [inline ui/app/templates/vault/cluster.hbs:11]: It might be useful for these selectors to be specific to customMessages since they'll likely show at the same time as other ui alert banners
```suggestion
      data-test-custom-alert={{bannerMessage.id}}

File: ui/lib/config-ui/addon/components/messages/page/create-and-edit-message-form.hbs [hunk 17]
```diff
@@ -18,43 +18,9 @@
     <MessageError @errorMessage={{this.errorBanner}} class="has-top-margin-s" />
 
     {{#each @message.formFields as |attr|}}
-      {{#if (eq attr.name "linkTitle")}}
-        <div class="field has-bottom-margin-m">
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/create-and-edit-message-form.hbs:22]: 👏

File: ui/lib/config-ui/addon/components/messages/page/details.hbs [hunk 18]
```diff
@@ -43,6 +43,16 @@
         (date-format (get @message attr.name) "MMM d, yyyy hh:mm aaa" withTimeZone=true)
       }}
     />
+  {{else if (eq attr.name "link")}}
+    {{#if (is-empty-value @message.link)}}
+      <InfoTableRow @label="Link" @value="None" />
+    {{else}}
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/details.hbs:49]: Could we remove this extra if/else and updated the block below to handle both cases?

```hbs
{{#each-in @message.link as |title href|}}
   <InfoTableRow @label="Link" @value={{or title "None"}}>
     {{#if @href}}
      <Hds::Link::Inline @icon="external-link" @href={{href}} data-test-link="message link">{{title}}</Hds::Link::Inline>
     {{/if}}
   </InfoTableRow>
{{/each-in}}

File: ui/lib/config-ui/addon/components/messages/page/list.hbs [hunk 19]
```diff
@@ -21,15 +21,15 @@
...
-      @type="add"
-      aria-label="message create"
-    >
-      Create message
-    </ToolbarLink>
+    <Hds::Button
+      @text="Create message"
+      @icon="plus"
+      @color="tertiary"
+      class="toolbar-link"
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/list.hbs:28]: Could you update these class to so they are consistent with the other HDS toolbar buttons we have?
```
@color="secondary"
class="toolbar-button"

File: ui/lib/config-ui/addon/components/messages/page/list.hbs [hunk 20]
```diff
@@ -115,4 +115,29 @@
...
+    <M.Header data-test-modal-title="maximum-message-modal">
+      Maximum number of messages reached
+    </M.Header>
+    <M.Body data-test-modal-body="maximum-message-modal">
+      Vault can only store up to 100 messages. To create a message, delete one of your messages to clear up space.
+    </M.Body>
+    <M.Footer>
+      <Hds::Button
+        @text="Confirm"
+        {{on "click" (fn (mut this.showMaxMessageModal) false)}}
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/list.hbs:138]: Mostly an FYI, hds modals can be closed by `{{on "click" F.close}}`

File: ui/lib/core/addon/components/kv-object-editor.hbs [hunk 21]
```diff
@@ -11,7 +11,7 @@
     class={{@labelClass}}
     data-test-kv-label={{true}}
   />
-  {{#if @validationError}}
+  {{#if (and @validationError (not @isSingleRow))}}
```
Comment (hellobontempo) [inline ui/lib/core/addon/components/kv-object-editor.hbs:14]: But don't you still need to validation error to display if the user attempts to just submit a title?

File: ui/app/models/config-ui/message.js [hunk 22]
```diff
@@ -11,10 +11,22 @@ import { withFormFields } from 'vault/decorators/model-form-fields';
 const validations = {
   title: [{ type: 'presence', message: 'Title is required.' }],
   message: [{ type: 'presence', message: 'Message is required.' }],
+  link: [
+    {
+      validator(model) {
+        if (!model?.link) return true;
+        const [linkTitle] = Object.keys(model.link);
+        const [linkHref] = Object.values(model.link);
```
[the remark this answers is not in the document]
Comment (hellobontempo) [reply ui/app/models/config-ui/message.js:19]: Also what happens if a user has a URL but doesn't want a title? Should we validate for both?

File: ui/lib/core/addon/components/kv-object-editor.js [hunk 23]
```diff
@@ -86,6 +90,9 @@ export default class KvObjectEditor extends Component {
   }
   @action
   validateKey(rowIndex, event) {
+    if (this.args.isSingleRow) {
```
Comment (hellobontempo) [inline ui/lib/core/addon/components/kv-object-editor.js:93]: I don't think validation should be reliant on `isSingleRow`, I think another arg should be passed that's called `allowWhiteSpace` or something

**Review by hellobontempo — COMMENTED** [review]
Some more comments! Great work on this, it's a lot of logic to sort through ⭐

File: ui/app/models/config-ui/message.js [hunk 24]
```diff
@@ -11,10 +11,22 @@ import { withFormFields } from 'vault/decorators/model-form-fields';
...
+        if (!linkTitle) return true;
+        return !!linkHref;
+      },
+      message: 'Link title and url is required.',
```
Comment (hellobontempo) [inline ui/app/models/config-ui/message.js:23]:
```suggestion
      message: 'Link title and url are required.',

File: ui/lib/config-ui/addon/components/messages/page/details.hbs
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/details.hbs:1]: I wonder if users will want to be able to see a preview of messages already created. Is the only way to see a preview during creation @yongqikuang?

File: ui/lib/config-ui/addon/components/messages/page/list.hbs [hunk 25]
```diff
@@ -115,4 +115,34 @@
       data-test-action-text="Create message"
     />
   </EmptyState>
+{{/if}}
+
+{{#if this.showMaxMessageModal}}
+  <Hds::Modal
+    id="maximum-message-modal"
+    @size="large"
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/list.hbs:123]: This should probably just be the standard modal size (I think you can remove this, because the default is medium)

File: ui/app/models/config-ui/message.js [hunk 26]
```diff
@@ -81,11 +93,14 @@ export default class MessageModel extends Model {
...
 
-  // the api returns link as an object with title and href as keys, but we separate the link key/values into
-  // different attributes to easily show link title and href fields on the create form. In our serializer,
-  // we send the link attribute in to the correct format (as an object) to the server.
-  @attr('string') linkTitle;
-  @attr('string') linkHref;
+  @attr('object', {
+    editType: 'kv',
+    keyPlaceholder: 'Display text (e.g. Learn more)',
+    valuePlaceholder: 'Paste URL (e.g. www.learnmore.com)',
```
[the remark this answers is not in the document]
Comment (kiannaquach) [reply ui/app/models/config-ui/message.js:99]: LOL 100% agree! Updating this!

File: ui/lib/config-ui/addon/components/messages/page/list.hbs [hunk 27]
```diff
@@ -115,4 +115,34 @@
...
+    <M.Body data-test-modal-body="maximum-message-modal">
+      Vault can only store up to 100 messages. To create a message, delete one of your messages to clear up space.
+    </M.Body>
+    <M.Footer>
+      <Hds::Button
+        @text="Confirm"
+        {{on "click" (fn (mut this.showMaxMessageModal) false)}}
+        data-test-modal-button="maximum-message-modal"
+      />
+      <Hds::Button
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/list.hbs:141]: Since both buttons do the same thing I don't think we need both. Elsewhere in the app we just render a single `Close` button for these cases
Screenshot 2024-01-19 at 12 47 33 PM

File: ui/app/templates/vault/cluster.hbs [hunk 28]
```diff
@@ -3,37 +3,45 @@
...
+      @onDismiss={{fn this.customMessages.onBannerDismiss bannerMessage.id}}
+      as |A|
+    >
+      <A.Title data-test-alert-title={{bannerMessage.id}}>{{bannerMessage.title}}</A.Title>
+      <A.Description data-test-alert-description={{bannerMessage.id}}>
+        {{bannerMessage.message}}
+        {{#unless (is-empty-value bannerMessage.link)}}
+          {{#each-in bannerMessage.link as |title href|}}
+            <Hds::Link::Inline @icon="external-link" @href={{href}}>{{title}}</Hds::Link::Inline>
```
Comment (hellobontempo) [inline ui/app/templates/vault/cluster.hbs:20]: I think you may way to add `@isHrefExternal={{true}}` to these

File: ui/app/templates/vault/cluster.hbs [hunk 29]
```diff
@@ -3,37 +3,45 @@
...
+{{#each this.customMessages.modalMessages as |modalMessage|}}
+  <Hds::Modal id={{modalMessage.id}} @size="large" @color="warning" data-test-modal={{modalMessage.id}} as |M|>
+    <M.Header data-test-modal-title={{modalMessage.id}}>
+      {{modalMessage.title}}
+    </M.Header>
+    <M.Body data-test-modal-body={{modalMessage.id}}>
+      {{modalMessage.message}}
+      {{#unless (is-empty-value modalMessage.link)}}
+        {{#each-in modalMessage.link as |title href|}}
+          <Hds::Link::Inline @icon="external-link" @href={{href}}>{{title}}</Hds::Link::Inline>
```
Comment (hellobontempo) [inline ui/app/templates/vault/cluster.hbs:36]:
```suggestion
          <Hds::Link::Inline @icon="external-link" @href={{href}} @isHrefExternal={{true}}>{{title}}</Hds::Link::Inline>

File: ui/app/templates/vault/cluster.hbs (hunk 15 above)
Comment (kiannaquach) [reply ui/app/templates/vault/cluster.hbs:8]: I confirmed with design and there isn't spacing that separates each banner :) This follows the same ui as cloud

File: ui/lib/config-ui/addon/components/messages/page/details.hbs [hunk 30]
```diff
@@ -43,6 +43,16 @@
...
+      {{#each-in @message.link as |title href|}}
+        <InfoTableRow @label="Link" @value={{title}}>
+          <Hds::Link::Inline @icon="external-link" @href={{href}} data-test-link="message link">{{title}}</Hds::Link::Inline>
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/details.hbs:52]: same here about `isHrefExternal`
```suggestion
          <Hds::Link::Inline @icon="external-link" @href={{href}} @isHrefExternal={{true}}>{{title}}</Hds::Link::Inline>
```

could you use the built in InfoTableRow selector combined with the `a` tag to assert this value instead of adding another select?

File: ui/lib/config-ui/addon/components/messages/page/details.hbs
Comment (kiannaquach) [reply ui/lib/config-ui/addon/components/messages/page/details.hbs:1]: Right now, users can always preview a message during creation and after creation

File: ui/lib/core/addon/components/kv-object-editor.hbs (hunk 21 above)
Comment (kiannaquach) [reply ui/lib/core/addon/components/kv-object-editor.hbs:14]: Screenshot 2024-01-19 at 1 12 56 PM
This is how it looks if we have both validations shown. The form field has an inline error message as well as the validationError in the kv-object-editor

File: ui/app/models/config-ui/message.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (kiannaquach) [reply ui/app/models/config-ui/message.js]: `const [title, href] = Object.entries(model.link)[0]` doesn't work when the object is `{"": ""}` since `Object.entries(model.link)` is an empty array. I did however add this suggested change 
```
// if there is a title or href, validate both have values otherwise return true return (title || href) ? title && href : true
```

File: ui/app/serializers/config-ui/message.js (hunk 11 above)
Comment (kiannaquach) [reply ui/app/serializers/config-ui/message.js]: Great idea to add to the transform :D

File: ui/lib/core/addon/components/kv-object-editor.hbs (hunk 21 above)
Comment (hellobontempo) [reply ui/lib/core/addon/components/kv-object-editor.hbs]: I still don't think this should be hidden based on whether or not it's a singleRow, that conditional isn't relevant. The issue here is both components manage validation. 

Might be worth asking the team, but the validation above the input is anti-pattern...so only one component here should be managing validation. I think the block should be removed all together (since it's also in the wrong place) but that could break a bunch of tests.

File: ui/app/models/config-ui/message.js [hunk 31]
```diff
@@ -11,10 +11,22 @@ import { withFormFields } from 'vault/decorators/model-form-fields';
 const validations = {
   title: [{ type: 'presence', message: 'Title is required.' }],
   message: [{ type: 'presence', message: 'Message is required.' }],
+  link: [
+    {
+      validator(model) {
+        if (!model?.link) return true;
+        const [title] = Object.keys(model.link);
+        const [href] = Object.values(model.link);
+        if (!title && !href) return true;
```
Comment (hellobontempo) [inline ui/app/models/config-ui/message.js:20]: This line is unnecessary because if there are neither then `title || href` will be falsy which returns true 
Screenshot 2024-01-19 at 1 39 22 PM
And you I think we should wrap `!!(title && href)` so it always returns a boolean

File: ui/app/transforms/object.js [hunk 32]
```diff
@@ -20,7 +20,7 @@ export default class ObjectTransform extends Transform {
     if (typeOf(value) !== 'object') {
       return {};
     } else {
-      return value;
+      return JSON.stringify(value) === '{"":""}' ? {} : value;
```
Comment (hellobontempo) [inline ui/app/transforms/object.js:23]: I like this! Hopefully it doesn't break things 🤞

File: ui/lib/config-ui/addon/components/messages/page/details.hbs [hunk 33]
```diff
@@ -43,6 +43,22 @@
         (date-format (get @message attr.name) "MMM d, yyyy hh:mm aaa" withTimeZone=true)
       }}
     />
+  {{else if (eq attr.name "link")}}
+    {{#if (is-empty-value @message.link)}}
+      <InfoTableRow @label="Link" @value="None" />
+    {{else}}
```
[the remark this answers is not in the document]
Comment (hellobontempo) [reply ui/lib/config-ui/addon/components/messages/page/details.hbs:49]: If I got that logic wrong, then I'd just revert the change all together, I was just trying to remove extra nested if/else conditionals

File: ui/lib/core/addon/components/kv-object-editor.js [hunk 34]
```diff
@@ -55,7 +56,10 @@ export default class KvObjectEditor extends Component {
   @action
   createKvData(elem, [value]) {
     this.kvData = KVObject.create({ content: [] }).fromJSON(value);
-    this.addRow();
+
```
Comment (hellobontempo) [inline ui/lib/core/addon/components/kv-object-editor.js:59]: What about just returning if this.args.isSingleRow?
`if (this.args.isSingleRow) return`

File: ui/lib/core/addon/components/form-field.hbs [hunk 35]
```diff
@@ -137,6 +137,10 @@
       @helpText={{@attr.options.helpText}}
       @subText={{@attr.options.subText}}
       @onKeyUp={{this.handleKeyUp}}
+      @keyPlaceholder={{@attr.options.keyPlaceholder}}
+      @valuePlaceholder={{@attr.options.valuePlaceholder}}
+      @isSingleRow={{@attr.options.isSingleRow}}
+      @allowWhiteSpace={{@attr.options.allowWhiteSpace}}
```
Comment (hellobontempo) [inline ui/lib/core/addon/components/form-field.hbs:143]: I know it's a lot of args but I think this is the way to go!

File: ui/app/models/config-ui/message.js (hunk 31 above)
Comment (kiannaquach) [reply ui/app/models/config-ui/message.js:20]: haha that makes sense! Thanks for helping with this one 🙈

File: ui/lib/core/addon/components/kv-object-editor.js (hunk 34 above)
Comment (kiannaquach) [reply ui/lib/core/addon/components/kv-object-editor.js:59]: We still want to add a single row on initial render (if it doesn't exist already). For instance, when we're editing a message with existing kv values, we don't want to add a row, however, if we're creating a new message, we want to add a row. Let me know if that makes sense!

File: ui/lib/core/addon/components/kv-object-editor.hbs (hunk 21 above)
Comment (kiannaquach) [reply ui/lib/core/addon/components/kv-object-editor.hbs]: Sounds good! I'll keep the validations here for now :)

Comment (hellobontempo) [reply ui/lib/core/addon/components/kv-object-editor.hbs]: We should definitely remove one of them - I think youre just the first to come across this bug!

File: ui/lib/core/addon/components/kv-object-editor.js (hunk 34 above)
Comment (hellobontempo) [reply ui/lib/core/addon/components/kv-object-editor.js:59]: Ah I see. I think the reverse logic is a little easier to follow `if (this.args.isSingleRow && value) return`, but either works! I'll defer to you. Feel free to leave as is :)

File: ui/lib/config-ui/addon/components/messages/page/details.hbs [hunk 36]
```diff
@@ -43,6 +43,14 @@
         (date-format (get @message attr.name) "MMM d, yyyy hh:mm aaa" withTimeZone=true)
       }}
     />
+  {{else if (eq attr.name "link")}}
+    {{#each-in @message.link as |title href|}}
+      <InfoTableRow @label="Link" @value={{or title "None"}}>
```
Comment (hellobontempo) [inline ui/lib/config-ui/addon/components/messages/page/details.hbs:48]: I forgot you're yielding a block here so I don't think this will render "None" if the block is empty 😭

**Review by hellobontempo — APPROVED** [review]
Great work on this! 🎉

File: ui/lib/config-ui/addon/components/messages/page/details.hbs (hunk 36 above)
Comment (kiannaquach) [reply ui/lib/config-ui/addon/components/messages/page/details.hbs]: Ahh you're right! Putting the old conditional logic back :)
